From: Emanuele Rocca <emanuele.rocca@arm.com>
Subject: coredump: add core_pattern specifier for si_code
Date: Tue, 17 Mar 2026 16:00:06 +0100

The specifiers supported by core_pattern include the option to indicate
the signal number si_signo by using %s.  Other than identifying which
signal generated a core dump (eg: 11 for SIGSEGV), it is useful to know
the reason why a certain signal was sent.  The signal code si_code (eg: 2
for SEGV_ACCERR) provides this information.

Adding the signal code to core_pattern can benefit in particular sysadmins
who pipe core dumps to user-space programs for later analysis. 
systemd-coredump(8) is a notable example of such programs.

Link: https://lore.kernel.org/ablsdmLsMKm0z5wt@NH27D9T0LF
Signed-off-by: Emanuele Rocca <emanuele.rocca@arm.com>
Cc: Alexander Viro <viro@zeniv.linux.org.uk>
Cc: Al Viro <viro@zeniv.linux.org.uk>
Cc: Christian Brauner <brauner@kernel.org>
Cc: Jan Kara <jack@suse.cz>
Cc: Jonathan Corbet <corbet@lwn.net>
Cc: Mark Brown <broonie@kernel.org>
Cc: Shuah Khan <skhan@linuxfoundation.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 Documentation/admin-guide/sysctl/kernel.rst |    1 +
 fs/coredump.c                               |    5 +++++
 2 files changed, 6 insertions(+)

--- a/Documentation/admin-guide/sysctl/kernel.rst~coredump-add-core_pattern-specifier-for-si_code
+++ a/Documentation/admin-guide/sysctl/kernel.rst
@@ -170,6 +170,7 @@ core_pattern
 	%d		dump mode, matches ``PR_SET_DUMPABLE`` and
 			``/proc/sys/fs/suid_dumpable``
 	%s		signal number
+	%n		signal code
 	%t		UNIX time of dump
 	%h		hostname
 	%e		executable filename (may be shortened, could be changed by prctl etc)
--- a/fs/coredump.c~coredump-add-core_pattern-specifier-for-si_code
+++ a/fs/coredump.c
@@ -403,6 +403,11 @@ static bool coredump_parse(struct core_n
 				err = cn_printf(cn, "%d",
 						cprm->siginfo->si_signo);
 				break;
+			/* code of the signal that caused the coredump */
+			case 'n':
+				err = cn_printf(cn, "%d",
+						cprm->siginfo->si_code);
+				break;
 			/* UNIX time of coredump */
 			case 't': {
 				time64_t time;
_
